Add recipe for jet

ericdallo opened the pull request:
### Brief summary of what the package does

Run [jet Clojure tool](https://github.com/borkdude/jet) from emacs for quickly convert between edn,json,yaml,transit.

### Direct link to the package repository

https://github.com/ericdallo/jet.el

### Your association with the package

Maintainer

### Relevant communications with the upstream package maintainer

None needed

### Checklist

<!-- Please confirm by replacing `[]` with `[x]`: -->

- x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] `M-x checkdoc` is happy with my docstrings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

<!-- After submitting, please fix any problems the CI reports. -->

progfolio commented:
Suggestions:

```emacs-lisp
(eval-when-compile (require 'subr-x))
```

Prevents the compiled code from loading the library.

```emacs-lisp
(defun jet--run (command &rest args)
  "Run COMMAND with ARGS."
  (jet--assert-jet-on-path)
  (with-output-to-string
    (with-current-buffer standard-output
      (shell-command (string-join (append (list command) args) " ") t jet-error-buffer-name))))
```

standard-output is not guaranteed to be a buffer. See the standard-output documentation for accepted values.

```emacs-lisp
(defun jet--major-mode-fn-for (to)
  "Retunrn the major mode function for TO."
  (or (cond ((string= "edn" to) (and (functionp 'clojure-mode) #'clojure-mode))
            ((string= "json" to) (and (functionp 'json-mode) #'json-mode))
            ((string= "yaml" to) (and (functionp 'yaml-mode) #'yaml-mode))
            ((string= "transit" to) nil)
            (t (and (functionp 'clojure-mode) #'clojure-mode)))
      (lambda ())))
```

Perhaps this would be better served by an alist? e.g.

```emacs-lisp
(defun jet--major-mode-fn-for (to)
  "Retunrn the major mode function for TO."
  (alist-get to jet-major-mode-alist #'ignore nil #'string=))
```

That would also allow for more easily customizing/adding to the options if necessary.

```emacs-lisp
(defun jet-menu--suffixes->args (suffixes)
  "Transform SUFFIXES in a single string."
  (string-join
   (thread-last
     suffixes
     (seq-filter (lambda (obj)
                   (cl-typep obj 'transient-option)))
     (seq-map (lambda (obj)
                (when-let (arg (oref obj value))
                  (concat (oref obj argument) (oref obj value))))))
   " "))
```

Can leverage when-let to obviate the need for thread-last:

```emacs-lisp
(defun jet-menu--suffixes->args (suffixes)
  "Transform SUFFIXES in a single string."
  (string-join
   (delq nil (seq-map (lambda (obj)
                        (when-let (((cl-typep obj 'transient-option))
                                   (arg (oref obj value)))
                          (concat (oref obj argument) arg)))
                      suffixes))
   " "))
```

`jet-menu--transient-suffix-by-name` unnecessarily uses the thread-last macro with a single form. It's clearer to just pass the suffixes directly as the second arg to `seq-find`.

Instead of building the command by concatenating user input, you could use `call-process` directly with the executable. That avoids the need to escape user input (which should be done if you decide to stick with using `shell-command`).

Consider adding autoload cookies to public commands. e.g. jet-print, jet-paste-cursor, jet-paste-buffer, etc.

checkdoc reports the following warnings:

> jet.el 35 41 note e-f-c Lisp symbol ‘jet-menu’ should appear in quotes jet.el 93 42 note e-f-c Lisp symbol ‘jet-menu’ should appear in quotes jet.el 111 24 note e-f-c Lisp symbol ‘jet-menu&#x2013;run’ should appear in quotes

ericdallo commented:
@progfolio thanks for the suggestions, I made all but the shell-command change since I couldn't find a way pipe the output and use in another command since most commands will be something like `echo '{:a 1 :b 2}' | jet `

> standard-output is not guaranteed to be a buffer. See the standard-output documentation for accepted values.

Is that really a issue? I just brought that from `shell-command-to-string` that does exactly that,I just passed the error buffer as additional param

progfolio commented:
> Is that really a issue? I just brought that from shell-command-to-string that does exactly that,I just passed the error buffer as additional param

Apologies, I glanced over the fact that you are using `with-output-to-string` which binds standard-output. No change necessary.

> I couldn't find a way pipe the output and use in another command since most commands will be something like echo '{:a 1 :b 2}' | jet

This seems like a shortcoming of the jet tool. I don't see why the application couldn't be made to interpret anything after `--` as immediate input. Might be worth opening a feature request. One way to work around the present design be to use `make-process`. That will start a async subprocess, which you can then send the input string to, block until the process is finished and collect the results. e.g.

```emacs-lisp
(defun jet--command (input &rest args)
  "Send INPUT to jet with ARGS.
Return jet output."
  (let ((p (make-process
            :name "jet"
            :command (append (list (executable-find "jet")) args)
            :filter (lambda (p output)
                      (process-put p :output (concat (process-get p :output) output)))
            :connection-type 'pipe)))
    (process-send-string p input)
    (process-send-eof p)
    (accept-process-output p)
    (process-get p :output)))

(jet--command "{:a 1 :b 2 :c 3}" "--to" "json")
```

    {
      "a" : 1,
      "b" : 2,
      "c" : 3
    }

If you do decide to continue using shell-command\* functions, be sure to escape any user input via `shell-quote-argument` so it does not get misinterpreted by their shell. Quoting is not necessary with the strategy I've outlined above.

ericdallo commented:
Thank you @progfolio, I will stick with shell-command for the simplicity, I updated to use shell-quote-argument so I think everything is good now

progfolio commented:
the byte compiler still has a few complaints:

```
jet.el                       34   1 warning  e-f-b-c  custom-declare-variable `jet-menu-max-length-to-echo' docstring has wrong usage of unescaped single quotes (use \=' or different quoting such as `...')
jet.el                       92   1 warning  e-f-b-c  docstring has wrong usage of unescaped single quotes (use \=' or different quoting such as `...')
jet.el                      110   1 warning  e-f-b-c  docstring has wrong usage of unescaped single quotes (use \=' or different quoting such as `...')

```

ericdallo commented:
Thanks, fixed, I wonder why my emacs doesn't catch that running `checkdox` via emacs 28

progfolio commented:
> Thanks, fixed, I wonder why my emacs doesn't catch that running `checkdoc` via emacs 28

Both checkdoc and the byte-compiler are developed as part of Emacs. I'm running Emacs 30, which includes the latest developments in both.

riscy commented:
Thanks both.  Wow, @progfolio -- Emacs 30?!  You're lightyears ahead of me.  This is a small diversion, but I wanted to ask about the suggestion you made:

> `(eval-when-compile (require 'subr-x))` / Prevents the compiled code from loading the library.

I think using `eval-when-compile` is the correct thing to do when the user is only making use of the macros in subr-x (e.g. if-let, when-let) but in some cases (e.g. `string-join` which isn't autoloaded) I would have thought it makes sense to bring in `subr-x`.

@ericdallo, here's a very quick first pass from my side:


`jet.el` with byte-compile using Emacs 28.1:
```
jet.el:210:10: Warning: the function `replace-region-contents' might not be
    defined at runtime.
```
This may be a situation like the one I described above, where it makes sense to bring `subr-x` completely.  Otherwise users will see byte-compiler warnings, at least on Emacs 28.1.

Also, we should add *formal* license boilerplate or an [SPDX-License-Identifier](https://spdx.dev/ids/) to jet.el -- although this package is GPL licensed via the GitHub repository, the file has no boilerplate or SPDX license.

Lastly, it looks like the description in your `;;; Commentary` could be expanded -- simply copying the summary line doesn't add much.  Common entry points, a link to the actual `jet` tool, etc. could help users make better decisions about using this package.

ericdallo commented:
Thank you for the review @riscy, last time I opened a melpa PR was 3 years ago and the review quality was already good IMO, it's nice to see how code quality evolved even further :)

I added the SPDX license comment, improved the commentary section and changed to use the require of subr-x, LMK if anything more.
I configured [CI](https://github.com/ericdallo/jet.el/actions/runs/4039463692/jobs/6944220893) to run lint/checkdoc/package-lint for multiple emacs version to avoid wasting your time, but it seems it's not working perfectly 😅

progfolio commented:
Chris Rayner ***@***.***> writes:

> Thanks both. Wow, @progfolio -- Emacs 30?! You're lightyears 
> ahead of me.
>This is a small diversion, but I wanted to ask about the 
>suggestion you made:

Hahaha. Yes. I rebuild off the master branch weekly.
Surprisingly stable and I enjoy testing/ironing out any bugs I run 
into.
 
>  (eval-when-compile (require 'subr-x)) / Prevents the compiled 
>  code from loading the
>  library.
>
> I think using eval-when-compile is the correct thing to do when 
> the user is only making use of
> the macros in subr-x (e.g. if-let, when-let) but in some cases 
> (e.g. string-join which isn't
> autoloaded) I would have thought it makes sense to bring in 
> subr-x.

It looks like string-join has been autoloaded more recently (Emacs 
28.2):

https://github.com/emacs-mirror/emacs/commit/84801d468ae1e9fed271151a9c7e259cf05488e8

I think the body of the eval-when-compile will evaluate the 
require at compile time. 
Because string-join is a defsubst, its definition is inlined in 
the resulting compiled code.
When interpreted, the eval-when-compile is essentially treated as 
a progn.
Perhaps I misunderstood, though.

I think you're right about replace-region-contents, which I 
overlooked.
Interestingly, I don't see that warning on my end unless the 
require statements are re-ordered:

```emacs-lisp
(eval-when-compile (require 'subr-x)) ;;byte compiler warns about 
replace-region-contents
(require 'transient)
```
vs

```emacs-lisp
(require 'transient)  ;; transient eval-when-compile's subr-x.
(eval-when-compile (require 'subr-x)) ;;Transitive dependency 
hiding warning?
```

riscy commented:
Thank you both, and also for the kind words @ericdallo. :)

Everything looks good here - merging.

ericdallo commented:
Thank you!
